mono/metadata/assembly-load-context.c

```diff
@@ -15,14 +16,79 @@ mono_alc_init (MonoAssemblyLoadContext *alc, MonoDomain *domain)
 	mono_loaded_images_init (li, alc);
 	alc->domain = domain;
 	alc->loaded_images = li;
+	alc->loaded_assemblies = NULL;
+	mono_coop_mutex_init (&alc->assemblies_lock);
 }
 
 void
 mono_alc_cleanup (MonoAssemblyLoadContext *alc)
 {
+	// TODO: get someone familiar with the GC to look at this
+	GSList *tmp;
+	MonoDomain *domain = alc->domain;
+
 	mono_loaded_images_free (alc->loaded_images);
+
+	for (tmp = alc->loaded_assemblies; tmp; tmp = tmp->next) {
```

Comment:
@CoffeeFlux and I talked offline: this code in `mono_alc_cleanup` isn't ready yet.  We're going to document our assumptions about the refcount on each assembly in the ALC and add a `g_assert_not_reached` for now and revisit once we get started on collectible ALC support after we grok the CoreCLR collectible assemblies design.